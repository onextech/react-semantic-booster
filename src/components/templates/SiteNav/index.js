import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { kebabCase } from 'lodash';
import { Sidebar, Dropdown, Responsive, Dimmer, Image } from 'semantic-ui-react';
import MenuLink from '../../atoms/MenuLink';
import DropdownLink from '../../atoms/DropdownLink';
import ButtonLink from '../../atoms/ButtonLink';
import Menu from '../../atoms/Menu';
import { mediaCssBreakpoints } from '../../../utils/responsive';


const SidebarPushable = styled(Sidebar.Pushable)`
  /* Button */
  .ui.menu.vertical .ui.button {
    width: 100%;
  }
  
  /* Fix for button padding in SUI menus */
  .ui.menu:not(.vertical) .item>.button.custom {
    padding-top: 0;
    padding-bottom: 0;
  }
  
  /* Dropdown */
  .ui.menu .ui.dropdown .menu>.item.fitted.link {
    display: flex;
    padding: 0 !important;
    a {
      color: inherit;
      &:focus, &:hover {
        text-decoration: none;
      }
    }
  }
  
  // Fix for dropdown link in  menu
  .inverted.ui.menu .ui.dropdown .menu > .item.fitted.link a {
    color: rgba(0,0,0,.87);
  }
  
  // Fix for dropdown text alignment in vertical menu
  .ui.menu.vertical .ui.item.dropdown {
    text-align: left;
  }
  
  /* Fix for dropdown menu not showing in sidebar due to SUI */
  .ui.sidebar {
    overflow-y: initial !important;
  }
  
  /* Sidebar */
  // Fix for menu items not positioned correctly
  .ui.secondary.pointing.menu .item {
    align-self: center;
  }
    
  // Fix for removing padding-left on hamburger icon in secondary menu only only
  .ui.secondary.pointing.menu:not(.sidebar) {
    .ui.item.dropdown[role="listbox"] {
        padding-left: 0;
    }
    .right, .left {
      .ui.item.dropdown[role="listbox"] {
        padding-left: 1.14286em;
      }
    }
  }
`;


class SiteNav extends React.Component {
  state = {}

  toggleSidebar = () => this.setState({ showSidebar: !this.state.showSidebar })

  handleDimmerClick = () => this.setState({ showSidebar: false });

  renderMenuItems = (items, isMobile = false) => {
    const result = [];

    const defaultMenuLinkProps = {};
    if (isMobile) {
      defaultMenuLinkProps.onClick = this.toggleSidebar;
    }

    items.map((item) => {
      // default menu item type
      let jsx = (
        <MenuLink
          to={item.to || '/'}
          key={kebabCase(item.name)}
          {...defaultMenuLinkProps}>
          {item.name}
        </MenuLink>
      );
      // render other menu item types
      switch (true) {
        case Boolean(item.image):
          jsx = (
            <MenuLink
              to={item.to}
              key={kebabCase(item.name)}
              {...defaultMenuLinkProps}>
              <Image {...item.image} />
            </MenuLink>
          );
          break;
        case Boolean(item.button):
          jsx = (
            <Menu.Item key={kebabCase(item.name)}>
              <ButtonLink
                to={item.to}
                content={item.name}
                {...item.button}
                {...defaultMenuLinkProps} />
            </Menu.Item>
          );
          break;
        case Boolean(item.dropdown):
          jsx = (
            <Dropdown item text={item.name} key={kebabCase(item.name)}>
              <Dropdown.Menu>
                {
                  item.dropdown.items.map(dropdownItem => (
                    <DropdownLink
                      to={dropdownItem.to}
                      key={kebabCase(dropdownItem.name)}
                      {...defaultMenuLinkProps}>
                      {dropdownItem.name}
                    </DropdownLink>
                  ))
                }
              </Dropdown.Menu>
            </Dropdown>
          );
          break;
        default:
          break;
      }
      return result.push(jsx);
    });
    return result;
  }

  renderDesktopMenu = () => {
    const { menuProps, menu } = this.props;
    return (
      <Menu attached {...menuProps}>
        <Responsive maxWidth={mediaCssBreakpoints.sm} as={Menu.Menu}>
          <Dropdown item icon="content" onClick={this.toggleSidebar} />
        </Responsive>
        {menu.map((submenu, i) => (
          <Responsive
            key={i}
            as={Menu.Menu}
            position={submenu.position}
            minWidth={mediaCssBreakpoints.sm}>
            {this.renderMenuItems(submenu.content)}
          </Responsive>
        ))}
      </Menu>
    );
  }

  renderMobileMenu = () => {
    const { menu } = this.props;
    const allMenus = [];
    if (menu.length > 1) {
      // merge menu contents
      menu.map(submenu => allMenus.push(...submenu.content));
    }
    return this.renderMenuItems(allMenus, true);
  }

  render() {
    const { children } = this.props;
    const { showSidebar } = this.state;

    return (
      <SidebarPushable>
        <Sidebar
          as={Menu}
          animation='push'
          width='thin'
          visible={showSidebar}
          icon='labeled'
          vertical
          inverted>
          {this.renderMobileMenu()}
        </Sidebar>
        <Sidebar.Pusher>
          {this.renderDesktopMenu()}
          <Dimmer active={showSidebar} onClick={this.handleDimmerClick} />
          {children}
        </Sidebar.Pusher>
      </SidebarPushable>
    );
  }
}

SiteNav.propTypes = {
  menu: PropTypes.arrayOf(PropTypes.shape({
    position: PropTypes.string.isRequired,
    content: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string.isRequired,
      to: PropTypes.string,
      dropdown: PropTypes.shape({
        items: PropTypes.arrayOf(PropTypes.shape({
          name: PropTypes.string.isRequired,
          to: PropTypes.string.isRequired,
        })).isRequired,
      }),
      button: PropTypes.object,
    })).isRequired,
  })).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.string,
    PropTypes.array,
  ]),
  menuProps: PropTypes.object,
};

SiteNav.defaultProps = {
  menuProps: undefined,
};

export default SiteNav;
