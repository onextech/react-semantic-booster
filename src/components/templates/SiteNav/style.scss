/* Icons */
// Fix for icons always having a right margin from SUI
.ui.menu .item>i.icon {
  margin: 0;
}

/* Button */
.ui.menu.vertical .ui.button {
  width: 100%;
}

/* Fix for button padding in SUI menus */
.ui.menu:not(.vertical) .item>.button.custom {
  padding-top: 0;
  padding-bottom: 0;
}

/* Dropdown */
.ui.menu .ui.dropdown .menu>.item.fitted.link {
  display: flex;
  padding: 0 !important;
  a {
    color: inherit;
    &:focus, &:hover {
      text-decoration: none;
    }
  }
}

// Fix for dropdown link in  menu
.inverted.ui.menu .ui.dropdown .menu > .item.fitted.link a {
  color: rgba(0,0,0,.87);
}

// Fix for dropdown text alignment in vertical menu
.ui.menu.vertical .ui.item.dropdown {
  text-align: left;
}

/* Fix for dropdown menu not showing in sidebar due to SUI */
.ui.sidebar {
  overflow-y: initial !important;
}


/* Sidebar */
// Fix for menu items not positioned correctly
.ui.secondary.pointing.menu .item {
  align-self: center;
}
