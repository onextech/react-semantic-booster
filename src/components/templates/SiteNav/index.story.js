import React from 'react';
import { storiesOf } from '@storybook/react';
import { Container } from 'semantic-ui-react';
import Block from '../../atoms/Block';
import SiteNav from './index';


export const menu = [ // eslint-disable-line import/prefer-default-export
  {
    position: 'left',
    content: [
      {
        name: 'Logo',
        to: '/',
        image: {
          height: 50,
          src: 'https://s3-ap-southeast-1.amazonaws.com/storage.api.merrymaker.com/images/merrymaker_mascot_head_side.svg',
        },
      },
      { name: 'Home', to: '/home' },
      { name: 'About', to: '/about' },
    ],
  },
  {
    position: 'right',
    content: [
      {
        name: 'Support',
        dropdown: {
          items: [
            { name: 'Phone', to: '/phone' },
            { name: 'Email', to: '/email' },
          ],
        },
      },
      { name: 'Contact Us', to: '/contact', button: { primary: true } },
    ],
  },
];

export const menuProps = {
  inverted: false,
  pointing: false,
  secondary: false,
  // Custom props not defined in semantic-ui
  custom: {
    float: false,
    container: false, // only applied when float = true
  },
};

export const floatedMenuProps = {
  inverted: true,
  pointing: true,
  secondary: true,
  // Custom props not defined in semantic-ui
  custom: {
    float: true,
    container: true, // only applied when float = true
  },
};

const ExampleContent = () => (
  <Block attached inverted textAlign="center">
    <Container>
      <div style={{ padding: '8em 0' }}>
        <h1>Hello World</h1>
      </div>
    </Container>
  </Block>
);

storiesOf('SiteNav', module)
  .add('Default', () => (
    <SiteNav menu={menu}>
      <ExampleContent />
    </SiteNav>
  ))
  .add('Inverted', () => (
    <SiteNav menu={menu} menuProps={{ inverted: true }}>
      <ExampleContent />
    </SiteNav>
  ))
  .add('Floated', () => (
    <SiteNav menu={menu} menuProps={floatedMenuProps}>
      <ExampleContent />
    </SiteNav>
  ));
