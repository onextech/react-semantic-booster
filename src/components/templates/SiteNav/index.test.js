import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter as Router } from 'react-router-dom';
import { render } from 'enzyme';
import SiteNav from '.';
import { menu } from './index.story';


describe('<SiteNav>', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Router><SiteNav menu={menu}>Hello World</SiteNav></Router>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has sidebar', () => {
    const wrapper = render(<Router><SiteNav menu={menu}>Hello World</SiteNav></Router>);
    expect(wrapper.hasClass('pushable')).toBe(true);
    expect(wrapper.find('.sidebar').length).toBe(1);
  });
});
