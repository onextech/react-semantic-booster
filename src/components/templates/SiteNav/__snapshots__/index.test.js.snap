// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteNav> renders correctly 1`] = `
<div
  className="ui segment pushable"
>
  <div
    className="ui inverted vertical labeled icon ui push left thin sidebar menu"
  >
    <div
      className="link fitted item"
      onClick={[Function]}
      style={
        Object {
          "padding": 0,
        }
      }
    >
      <a
        className="sc-ifAKCX gPFwxH"
        custom={
          Object {
            "basic": true,
          }
        }
        href="/"
        onClick={[Function]}
      >
        <img
          className="ui image"
          height={50}
          src="https://s3-ap-southeast-1.amazonaws.com/storage.api.merrymaker.com/images/merrymaker_mascot_head_side.svg"
        />
      </a>
    </div>
    <div
      className="link fitted item"
      onClick={[Function]}
      style={
        Object {
          "padding": 0,
        }
      }
    >
      <a
        className="sc-EHOje fafhJU"
        custom={undefined}
        href="/home"
        onClick={[Function]}
      >
        Home
      </a>
    </div>
    <div
      className="link fitted item"
      onClick={[Function]}
      style={
        Object {
          "padding": 0,
        }
      }
    >
      <a
        className="sc-bZQynM hnksPx"
        custom={undefined}
        href="/about"
        onClick={[Function]}
      >
        About
      </a>
    </div>
    <div
      aria-busy={undefined}
      aria-disabled={undefined}
      aria-expanded={false}
      aria-multiselectable={undefined}
      className="ui item dropdown"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      role="listbox"
      tabIndex={0}
    >
      <div
        aria-live="polite"
        className="text"
        role="alert"
      >
        Support
      </div>
      <i
        aria-hidden="true"
        className="dropdown icon"
        onClick={[Function]}
      />
      <div
        className="menu transition"
      >
        <div
          aria-checked={undefined}
          aria-disabled={undefined}
          aria-selected={undefined}
          className="item fitted link"
          onClick={[Function]}
          role="option"
        >
          <a
            className="sc-bwzfXH cBTmMw"
            href="/phone"
            onClick={[Function]}
          >
            Phone
          </a>
        </div>
        <div
          aria-checked={undefined}
          aria-disabled={undefined}
          aria-selected={undefined}
          className="item fitted link"
          onClick={[Function]}
          role="option"
        >
          <a
            className="sc-bwzfXH cBTmMw"
            href="/email"
            onClick={[Function]}
          >
            Email
          </a>
        </div>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
    >
      <button
        className="ui primary button custom sc-htpNat cnXOeu"
        disabled={undefined}
        onClick={[Function]}
        role="button"
        tabIndex={undefined}
      >
        <a
          href="/contact"
          onClick={[Function]}
        >
          Contact Us
        </a>
      </button>
    </div>
  </div>
  <div
    className="pusher"
  >
    <div
      className="ui attached menu"
      custom={
        Object {
          "container": false,
          "float": false,
        }
      }
    >
      <div
        className="menu"
      >
        <div
          aria-busy={undefined}
          aria-disabled={undefined}
          aria-expanded={false}
          aria-multiselectable={undefined}
          className="ui item dropdown"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <div
            aria-live="polite"
            className="text"
            role="alert"
          />
          <i
            aria-hidden="true"
            className="content icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
          />
        </div>
      </div>
    </div>
    <div
      className="ui dimmer"
      onClick={[Function]}
    />
    Hello World
  </div>
</div>
`;
